src/dbnode/storage/bootstrap/bootstrapper/base.go

```diff
@@ -76,133 +74,261 @@ func (b baseBootstrapper) String() string {
 	return baseBootstrapperName
 }
 
-func (b baseBootstrapper) Can(strategy bootstrap.Strategy) bool {
-	return b.src.Can(strategy)
-}
-
-func (b baseBootstrapper) BootstrapData(
-	namespace namespace.Metadata,
-	shardsTimeRanges result.ShardTimeRanges,
-	opts bootstrap.RunOptions,
-) (result.DataBootstrapResult, error) {
-	if shardsTimeRanges.IsEmpty() {
-		return result.NewDataBootstrapResult(), nil
-	}
-	step := newBootstrapDataStep(namespace, b.src, b.next, opts)
-	err := b.runBootstrapStep(namespace, shardsTimeRanges, step)
-	if err != nil {
-		return nil, err
+func (b baseBootstrapper) Bootstrap(
+	namespaces bootstrap.Namespaces,
+) (bootstrap.NamespaceResults, error) {
+	logFields := []zapcore.Field{
+		zap.String("bootstrapper", b.name),
 	}
-	return step.result(), nil
-}
 
-func (b baseBootstrapper) BootstrapIndex(
-	namespace namespace.Metadata,
-	shardsTimeRanges result.ShardTimeRanges,
-	opts bootstrap.RunOptions,
-) (result.IndexBootstrapResult, error) {
-	if shardsTimeRanges.IsEmpty() {
-		return result.NewIndexBootstrapResult(), nil
-	}
-	step := newBootstrapIndexStep(namespace, b.src, b.next, opts)
-	err := b.runBootstrapStep(namespace, shardsTimeRanges, step)
-	if err != nil {
-		return nil, err
+	curr := bootstrap.Namespaces{
+		Namespaces: bootstrap.NewNamespacesMap(bootstrap.NamespacesMapOptions{}),
 	}
-	return step.result(), nil
-}
+	for _, elem := range namespaces.Namespaces.Iter() {
+		id := elem.Key()
+		namespace := elem.Value()
 
-func (b baseBootstrapper) runBootstrapStep(
-	namespace namespace.Metadata,
-	totalRanges result.ShardTimeRanges,
-	step bootstrapStep,
-) error {
-	prepareResult, err := step.prepare(totalRanges)
-	if err != nil {
-		return err
-	}
+		// Shallow copy the namespace, do not modify namespaces input to bootstrap call.
+		currNamespace := namespace
 
-	var (
-		wg                     sync.WaitGroup
-		currStatus, nextStatus bootstrapStepStatus
-		currErr, nextErr       error
-		nextAttempted          bool
-	)
-	currRanges := prepareResult.currAvailable
-	nextRanges := totalRanges.Copy()
-	nextRanges.Subtract(currRanges)
-	if !nextRanges.IsEmpty() &&
-		b.Can(bootstrap.BootstrapParallel) &&
-		b.next.Can(bootstrap.BootstrapParallel) {
-		// If ranges can be bootstrapped now from the next source then begin attempt now
-		nextAttempted = true
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			nextStatus, nextErr = step.runNextStep(nextRanges)
-		}()
-	}
+		dataRange := currNamespace.DataRunOptions.ShardTimeRanges.MinMaxRange()
```

Comment:
I think that this may be better so you don't have to iterate over the range twice using `MinMaxRange()`

```golang
dataMin, dataMax := currNamespace.DataRunOptions.ShardTimeRanges.MinMax()
dataRange := max.Sub(min)
```

Reply:
Also, using `namespace.DataRunOptions.ShardTimeRanges` and `namespace.DataRunOptions.RunOptions` pretty often throughout this function, may be better to stick them in vars